lib/refinement/changeset.rb

```diff
@@ -0,0 +1,211 @@
+require 'cocoapods/executable'
+require 'set'
+
+module Refinement
+  # Represents a set of changes in a repository between a prior revision and the current state
+  class Changeset
+    # An error that happens when computing a git diff
+    class GitError < Error; end
+    require 'refinement/changeset/file_modification'
+
+    # @return [Pathname] the path to the repository
+    attr_reader :repository
+    # @return [Array<FileModification>] the modfications in the changeset
+    attr_reader :modifications
+    # @return [Hash<Pathname,FileModification>] modifications keyed by relative path
+    attr_reader :modified_paths
+    # @return [Hash<Pathname,FileModification>] modifications keyed by relative path
+    attr_reader :modified_absolute_paths
+
+    private :modifications, :modified_paths, :modified_absolute_paths
+
+    def initialize(repository:, modifications:)
+      @repository = repository
+      @modifications = self.class.add_directories(modifications).uniq.freeze
+
+      @modified_paths = {}
+      @modifications
+        .each { |mod| @modified_paths[mod.path] = mod }
+        .each { |mod| @modified_paths[mod.prior_path] ||= mod if mod.prior_path }
+      @modified_paths.freeze
+
+      @modified_absolute_paths = {}
+      @modified_paths
+        .each { |path, mod| @modified_absolute_paths[path.expand_path(repository).freeze] = mod }
+      @modified_absolute_paths.freeze
+    end
+
+    # @visibility private
+    # @return [Array<FileModification>] file modifications that include modifications for each
+    #                                   directory that has had a child modified
+    # @param  modifications [Array<FileModification>] The modfications to add directory modifications to
+    def self.add_directories(modifications)
+      dirs = Set.new
+      add = lambda { |path|
+        break unless dirs.add?(path)
+        add[path.dirname]
+      }
+      modifications.each do |mod|
+        add[mod.path.dirname]
+        add[mod.prior_path.dirname] if mod.prior_path
+      end
+      modifications +
+        dirs.map { |d| FileModification.new(path: Pathname("#{d}/").freeze, type: FileModification::DIRECTORY_CHANGE_TYPE) }
+    end
+
+    # @return [FileModification,Nil] the changeset for the given absolute path,
+    #   or `nil` if the given path is un-modified
+    # @param absolute_path [Pathname]
+    def include_path?(absolute_path:)
```

Comment:
not sure whats the ruby-ism here but this method with `?` conveys `Boolean` return type rather than a `FileModification` return type.

Reply:
yeah agreed

Reply:
all of those have been renamed to be less awkward